core/src/main/java/org/mule/runtime/core/api/processor/MessageProcessors.java

```diff
@@ -108,15 +110,15 @@ public static InternalEvent processToApply(InternalEvent event, ReactiveProcesso
    * @return result event
    * @throws MuleException if an error occurs
    */
-  public static InternalEvent processToApply(InternalEvent event, ReactiveProcessor processor, boolean completeContext)
+  public static BaseEvent processToApply(BaseEvent event, ReactiveProcessor processor, boolean completeContext)
       throws MuleException {
     try {
       return just(event)
           .transform(processor)
           // Ensure errors handled by MessageProcessorChains are returned
-          .switchIfEmpty(from(event.getContext().getResponsePublisher()))
-          .doOnSuccess(completeSuccessIfNeeded(event.getContext(), completeContext))
-          .doOnError(completeErrorIfNeeded(event.getContext(), completeContext))
+          .switchIfEmpty(from(((BaseEventContext) event.getContext()).getResponsePublisher()))
```

Comment:
These casts to BaseEventContext are really ugly. One thing is to cast like this in the tests, but the production code depends on EventContext according to the API, so our API is wrong and must be extended or the code must manage the situation when it received a eventContext that is not a BaseEventContext.
This comment applies in many places on this PR

Reply:
Created https://www.mulesoft.org/jira/browse/MULE-13494 to do next